[processor/k8sattributes] Automatic resource attributes - annotation prefix

zeitlinger opened the pull request:
This is the first part of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37114 - the ability to use an annotation prefix to define a resource attribute

zeitlinger committed:
automatically set service instance id

zeitlinger committed:
automatically set service instance id

zeitlinger committed:
automatically set service instance id

zeitlinger committed:
automatically set service instance id

zeitlinger committed:
add label rules, test annotation override

zeitlinger committed:
add service version

zeitlinger committed:
automatically set service name

zeitlinger committed:
automatically set service name

zeitlinger committed:
automatically set service name

zeitlinger committed:
automatically set service name

zeitlinger committed:
automatically set service name

zeitlinger committed:
lint

zeitlinger committed:
add changelog entry

zeitlinger committed:
add docs

zeitlinger committed:
part-of is not part of semconv

zeitlinger committed:
add app.kubernetes.io/instance

zeitlinger committed:
update service.version

zeitlinger committed:
add app.kubernetes.io/instance

zeitlinger committed:
rename

zeitlinger committed:
fix merge, revert implicit "enabled: true"

zeitlinger committed:
custom prefixes

zeitlinger committed:
add, customizations and missing namespace rule

zeitlinger committed:
add, customizations and missing namespace rule

zeitlinger committed:
add, customizations and missing namespace rule

zeitlinger committed:
format

zeitlinger committed:
format

zeitlinger committed:
format

zeitlinger committed:
strip down to just annotation prefix for the first PR

zeitlinger committed:
strip down to just annotation prefix for the first PR

zeitlinger committed:
strip down to just annotation prefix for the first PR

mattsains reviewed (approved):
I am not an approver unfortunately, but I looked through this PR and it looks good to me

mattsains commented:
perhaps @andrzej-stencel could take a look?

dmitryax reviewed (changes requested):
Let's agree on the configuration interface before proceeding https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37114/files#r2025819122

zeitlinger commented:
@dmitryax answering here:

> Can we just move well known attributes from this section to where all the other pre-defined resource attributes are configured `extract.metadata`, and disabled by default?
> 
> Then this section would be responsible for taking resource attributes from annotations only. Instead of calling it `automatic_attributes`, it can be something like
> 
> ```
> extract:
>   otel_annotations: <bool> // false by default
> ```

will do

> 
> Then, if we want to provide flexibility for annotations extraction, we could add `extract.annotations` with the same syntax as we currently have for `extract.labels`. That is not really needed right now.

will remov

> 
> Providing such interface would be consistent with the existing one and easier to understand by the end users. Please let me know WDYT.

sounds good :smile:

zeitlinger committed:
use otel_annotations

zeitlinger commented:
@dmitryax done - the test failures seem unrelated to this PR

dmitryax commented on processor/k8sattributesprocessor/internal/kube/automatic.go:
```diff
@@ -0,0 +1,26 @@
+	"regexp"
+)
+
+type AutomaticRules struct {
+	Enabled            bool     `mapstructure:"enabled"`
+	AnnotationPrefixes []string `mapstructure:"annotation_prefixes"`
+}
```
This struct shouldn't be needed anymore

dmitryax commented on processor/k8sattributesprocessor/internal/kube/automatic.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+func AutomaticAnnotationRule(prefix string) FieldExtractionRule {
+	return FieldExtractionRule{
+		Name:                 "$1",
+		KeyRegex:             regexp.MustCompile(`^` + regexp.QuoteMeta(prefix) + `(.+)$`),
+		HasKeyRegexReference: true,
+		From:                 MetadataFromPod,
+	}
```
This can be just a variable. No need for a function and passing the const

```
var defaultOtelAnnotationExtractionRule = FieldExtractionRule{
		Name:                 "$1",
		KeyRegex:             regexp.MustCompile(`^` + regexp.QuoteMeta(defaultOtelAnnotationPrefix) + `(.+)$`),
		HasKeyRegexReference: true,
		From:                 MetadataFromPod,
	}
```

Can probably be moved to processor/k8sattributesprocessor/options.go or even right inside `withOtelAnnotations`.

dmitryax commented:
I actually missed that we already have `extract::annotations` config option. What if instead of adding another boolean flag `otel_annotations`, we just add this rule by default in `extract::annotations`. Users who don't want it, can disable it by setting it to an empty slice. The prefix is very specific, I don't think it's being used extensively, so it shouldn't surprise the users. WDYT?

zeitlinger committed:
remove unneeded methods

zeitlinger commented:
> I actually missed that we already have `extract::annotations` config option. What if instead of adding another boolean flag `otel_annotations`, we just add this rule by default in `extract::annotations`. Users who don't want it, can disable it by setting it to an empty slice. The prefix is very specific, I don't think it's being used extensively, so it shouldn't surprise the users. WDYT?
> 
> If that sounds good, we should clearly describe it in the changelog and how to disable it.

I like that :smile:

zeitlinger committed:
fix rebase

zeitlinger commented:
> > I actually missed that we already have `extract::annotations` config option. What if instead of adding another boolean flag `otel_annotations`, we just add this rule by default in `extract::annotations`. Users who don't want it, can disable it by setting it to an empty slice. The prefix is very specific, I don't think it's being used extensively, so it shouldn't surprise the users. WDYT?
> > If that sounds good, we should clearly describe it in the changelog and how to disable it.
> 
> I like that 😄

@dmitryax on second thought - what about users who do have their own attribute rules.

They would have to add a rule with the exact prefix - I would like to avoid that users have to know those details.

ChrsMark commented on processor/k8sattributesprocessor/README.md:
```diff
@@ -271,10 +271,12 @@ k8sattributes/2:
+        key: app.kubernetes.io/component
+        from: pod
+    # Apply the operator rules - see https://github.com/open-telemetry/semantic-conventions/blob/main/docs/non-normative/k8s-attributes.md
+    otel_annotations: true # default is false - use true to enable
```
I'm fine with the addition of this specific one. However, I wonder how this will be combined with the rest of the settings as described at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37114#discussion_r2035380275. Sorry if I miss any context here but that'd help clarifying this.

zeitlinger replied:
I think we'll need to have another 2 flag to cover this

ChrsMark replied:
Should we decide on those at a high level before adding this specific one? My preference would be to have alignment on the high level config schema before starting any partial implementations. Would that make sense?

dmitryax replied:
I don't think we need any other config options for other things. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37114/files#r2025819122

ChrsMark replied:
What about the well known labels that are described at https://github.com/open-telemetry/opentelemetry-operator?tab=readme-ov-file#configure-resource-attributes-with-labels and was originally added in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37114/files#diff-57b0d6448d516c0c80b326a7f7f95c3256b2b774e496870dc1dd35dcf8bfc256R281?

ChrsMark replied:
Hmm, `extract::metadata` is a map so I'm not sure how we could add extra options without breaking it, right?
Also in this PR we extract the annotation through the `extract::otel_annotations` setting so I assume the equivalent would be `extact::well_known_labels`, @zeitlinger could you confirm if that's the direction you had in mind?

ChrsMark replied:
From my perspective I'm coverd with the current state and the future plans. @dmitryax feel free to resolve this comment if we are on agreement.

ChrsMark reviewed (commented):
I like the idea of keeping the `extract:otel_annotations` setting different from the `extract::annotations`. Changing the behaviour of `extract::annotations` now to include the logic of this PR would affect users as a breaking change and disabling this or just overriding the rules would require extra manual work for the users. In addition, it's not a bad design to group this rulings under a specific setting as a shortcut, right? 

Based on this I'd lean towards going with the separate one.

zeitlinger commented on processor/k8sattributesprocessor/README.md:
```diff
@@ -271,10 +271,12 @@ k8sattributes/2:
+      - tag_name: app.label.component
+        key: app.kubernetes.io/component
+        from: pod
+    # Apply the operator rules - see https://github.com/open-telemetry/semantic-conventions/blob/main/docs/non-normative/k8s-attributes.md
```
no - operator rules was the old name

zeitlinger replied:
maybe `Apply the recommended attributes`

dmitryax reviewed (approved):
LGTM

dmitryax commented:
> They would have to add a rule with the exact prefix - I would like to avoid that users have to know those details.

I was thinking of avoiding complicating the configuration interface which is already big and complex, but this is a valid concern. So I agree with keeping another option.

dmitryax commented:
PTAL at the failing test

ChrsMark commented on processor/k8sattributesprocessor/config.go:
```diff
@@ -166,6 +166,10 @@ type ExtractConfig struct {
+
+	// OtelAnnotations extracts all pod annotations with the prefix "resource.opentelemetry.io" as resource attributes
+	// E.g. "resource.opentelemetry.io/foo" becomes "foo"
+	OtelAnnotations bool `mapstructure:"otel_annotations"`
```
Can we also add a few words of text in the docs' README because right now the setting is only hidden here and in the configuration example of the README. Having an explicit section for this would make sense.

ChrsMark reviewed (commented):
I think https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38825/files#r2035383252 still needs to be clarified :).

Overall looks good to me with a minor suggestion. Let's put a few words about this option in a section in the README as well to ensure users can spot and know about it not only from the example config.

zeitlinger committed:
Update processor/k8sattributesprocessor/README.md

Co-authored-by: Christos Markou <chrismarkou92@gmail.com>

zeitlinger commented:
@ChrsMark 

> Hmm, extract::metadata is a map so I'm not sure how we could add extra options without breaking it, right?

not sure what you mean by breaking? API compatibility?

> Also in this PR we extract the annotation through the extract::otel_annotations setting so I assume the equivalent would be extact::well_known_labels

originally I had in mind to have a grouping element for 
- annotations
- well known labels
- recommended attributes 

.. where you opt into annotations and recommended attributes with a single flag and well known attributes optionally after that.

```yaml
extract:
  recommeded: 
    enabled: true
    well_known_attributes: true
```

.. but I'm fine with having them top-level as well

See https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-attributes/#service-attributes

zeitlinger committed:
fix test

zeitlinger committed:
add readme section

zeitlinger commented:
> PTAL at the failing test

fixed - the remaining failure is unrelated and addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39300

ChrsMark commented on processor/k8sattributesprocessor/README.md:
```diff
@@ -251,6 +251,18 @@ extract:
+The processor can be configured to set the 
+[recommended resource attributes](https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-attributes/):
+
+- `otel_attributes` will translate `resource.opentelemetry.io/foo` to the `foo` resource attribute, etc.
```
Do we mean `otel_annotations`?

zeitlinger replied:
yes

zeitlinger commented:
> I'm fine with having this as top level setting and then having another top level setting for `well_knonw_labels`. So we will end up with the following:
> 
> ```yaml
> extract:
>   otel_annotations: true 
>   well_known_labels: true
> ```

We will need a third one for the [service attributes](https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-attributes/#service-attributes)

zeitlinger committed:
fix attribute

ChrsMark commented:
> We will need a third one for the [service attributes](https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-attributes/#service-attributes)

I see, my only concern would be that the `service attributes` group will have an overlap with the other 2 but we can discuss it then (probably that should be fine though).

I'm going to approve this PR as is since this top level `otel_annotations` setting makes sense on its own and the path forward seems to be clear.

cyrille-leclerc commented:
Should we evolve the OpenTelemetry Collector Helm Chart preset `presets/kubernetesAttributes` ([docs](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/opentelemetry-collector-0.131.0/charts/opentelemetry-collector#configuration-for-kubernetes-attributes-processor)) to activate by default `otel_annotations: true` ([docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor#configuring-recommended-resource-attributes)) and to extract `service.name`, `service.namespace`, `service.instance.id`,  `service.version`, and IMO `deployment.environment.name`?

See https://github.com/open-telemetry/opentelemetry-helm-charts/blob/opentelemetry-collector-0.131.0/charts/opentelemetry-collector/templates/_config.tpl#L286-L327
